# amnh-museumomics
Workshop on museum genomics for SCCS-NY 2022 at AMNH

## Workshop background ##

## Study background: changing chipmunks ##

Bi et al. 2015, Bi et al. 2019, Herrera et al. 2022

## Fetching publicly available data ##

### SRA ###

`hist_sra_numbers = ["SRR3172000", "SRR3172001", "SRR3172003", "SRR3172004", "SRR3172005", "SRR3172006", "SRR3172007", "SRR3172008", "SRR3172009", "SRR3172010", "SRR3172011", "SRR3172012", "SRR3172014", "SRR3172015", "SRR3172016", "SRR3172018", "SRR3172020", "SRR3172021", "SRR3172022", "SRR3172023", "SRR3172024", "SRR3172025", "SRR3172027", "SRR3172028", "SRR3172029", "SRR3172030", "SRR3172031", "SRR3172032", "SRR3172033", "SRR3172035", "SRR3172036", "SRR3172037", "SRR3172039", "SRR3172040", "SRR3172041", "SRR3172042", "SRR3172043", "SRR3172044", "SRR3172045", "SRR3172047", "SRR3172048", "SRR3172049", "SRR3172051", "SRR3172052", "SRR3172053", "SRR3172054", "SRR3172055", "SRR3172056", "SRR3172057", "SRR3172059", "SRR3172060", "SRR3172061"]`


### Reference genome ###

## Trimming and aligning data with Paleomix

https://paleomix.readthedocs.io/en/stable/

### Looking at Paleomix output ###

## Variant calling with BCFtools ##

## Variant filtering with SAMtools ##

## Getting started with R ##

## Importing VCF and converting to ##

## Running DAPC ##

## Running structure ## 

## Running an outlier test ##

## Warp up ##
